src/examples/algorithms/Anomaly.cpp

```diff
@@ -0,0 +1,199 @@
+    return dest;
+}
+
+
+// Converts an array of 1s and 0s to a sparse vector
+std::vector<UInt> binaryToSparse(const std::vector<UInt>& v)
+{
+    std::vector<UInt> sparse;
+    for (UInt index = 0; index < v.size(); index++) {
+        if (v[index] > 0) { // generic form, so values like 0.3 works as well
+            sparse.push_back(index);
+        }
+    }
+    return sparse;
+}
+
+
+// Converts the output of the temporal pooler (columns of 4 cells) to a single-cell-per-column
+// representation
+std::vector<UInt> normalizeTPOutput(const std::vector<UInt>& v, UInt cellsPerColumn)
+{
+    std::vector<UInt> normalized;
+    for (unsigned int i = 0; i < v.size(); ) {
+        unsigned int val = 0;
+        unsigned int j;
+        for (j = i; j < v.size() && j < i + cellsPerColumn; j++) {
+            val |= v[j];
+        }
+        normalized.push_back(val);
+        i = j;
+    }
+    return normalized;
+}
+
+// the chain of: input -> encoder -> spatial pooler -> temporal memory -> anomaly -> output
+int main()
+{
+    using namespace nupic;
+    using namespace nupic::algorithms::spatial_pooler;
+    using namespace nupic::algorithms::Cells4;
+    using namespace nupic::algorithms::anomaly;
+
+    std::string data("data.csv");
+    const UInt COLS = 2048;
+    const UInt CELLS = 4;
+
+    // Open the file or exit with error code 1
+    std::ifstream in(data.c_str());
+    if (!in.is_open()) return 1;
+
+    // Declare a vector for each column in the CSV file
+    std::vector<float> column1;
+
+    // Read the CSV file into a couple of vectors
+    std::string line;
+    while (getline(in, line)) {
+        // Use a boost tokenizer to parse a CSV line
+        boost::tokenizer<boost::escaped_list_separator<char>> tok(line);
+        std::vector<std::string> tokens(tok.begin(), tok.end());
+
+        // Skip line if it the aren't 2 elements at least.
+        if (std::distance(tok.begin(), tok.end()) < 1) continue;
+
+        // Store both items
+        column1.push_back(std::stof(tokens[0]));
+    }
+
+    std::cout << "Column1: ";
+    print_vector(column1, ' ');
+    std::cout << std::endl;
+
+    ScalarEncoder encoder{
+        25,   // w: Active bits per value
+        0,   // min value
+        10,  // max value
+        250,  // n: bits per output value,
+        0,
+        0,
+        true // Clip input to min/max vlues if needed.
+    };
+
+    // Spatial Pooler
+    SpatialPooler sp{
+        std::vector<UInt>{encoder.getOutputWidth()}, // Input dimensions = encoder width
+        std::vector<UInt>{COLS}  // Number of colums per dimension
+    };
+
+    // Temporal Pooler
+    Cells4 tp{
+        COLS,    // Number of columns
+        CELLS,     // Cells per column
+        12, 
+        8, 
+        15, 
+        5, 
+        .5, 
+        .8, 
+        1.0, 
+        .1, 
+        .1, 
+        0.0, 
+        false, 
+        42, 
+        true, 
+        false
```

Comment:
are the defaults updated to the optimal values? Or what is called the "getOptimalParams()" as discussed and implemented a while ago?